app/services/custom_fields/hierarchy/hierarchical_item_formatter.rb

```diff
@@ -0,0 +1,105 @@
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module CustomFields
+  module Hierarchy
+    class HierarchicalItemFormatter
+      include NumberFormatHelper
+
+      # @param [Boolean] path Show the full item ancestor path if true
+      # @param [Boolean] label Show the item label if true
+      # @param [Boolean] suffix Show the item short or weight if true
+      # @param [Boolean] suffix_parentheses Wrap suffix in parentheses.
+      # @param [Integer] number_length_limit If formatting numbers, restrict the whole length of the number to the
+      # given limit before falling back to scientific notation.
+      # @param [Integer] number_integer_digit_limit If formatting numbers, restrict the integer part of the number to
+      # the given limit before falling back to scientific notation.
+      # @param [Integer] number_precision If formatting numbers, round to the given precision.
+      def initialize(path: false,
+                     label: true,
+                     suffix: true,
+                     suffix_parentheses: true,
+                     number_length_limit: 9,
+                     number_integer_digit_limit: 8,
+                     number_precision: 4)
+        @path = path
+        @label = label
+        @suffix = suffix
+        @suffix_parentheses = suffix_parentheses
+        @number_length_limit = number_length_limit
+        @number_integer_digit_limit = number_integer_digit_limit
+        @number_precision = number_precision
+      end
+
+      def format(item:)
+        parts = []
+
+        if @path && @label
+          parts << branch(item)
+        elsif @path
+          parts << ancestors(item)
+        elsif @label
+          parts << item.label
+        end
+
+        if @suffix
+          suffix = format_suffix(item)
+          parts << suffix if suffix.present?
+        end
+
+        parts.join(" ")
+      end
+
+      private
+
+      def branch(item) = persistence_service.get_branch(item:).value!.filter_map(&:label).join(" / ")
+
+      def ancestors(item) = persistence_service.get_ancestors(item:).value!.filter_map(&:label).join(" / ")
+
+      def format_suffix(item)
+        return "" if item.short.nil? && item.weight.nil?
```

Comment:
Will short ever be nil for a hierarchy? Since users submit items through a form that has an input for the short, I'd expect them to submit an empty string and thus I'd expect short to always be a string that sometimes is `""`.

In other words: I'd probably rewrite this as

```suggestion
        return nil if item.short.blank? && item.weight.nil?
```

Reply:
So, in general, we should NEVER store an empty string to a short. That's what our contracts should guard us for. A short is either present or nil. And it is mutual exclusive with a weight, that is either a present or nil.

The only difference in the contracts is, that the list item is valid without a short (nil), while a weighted list item must have a weight.

Reply:
Indeed. I didn't expect this, but you are right. Then let's keep it as strict as you wrote it.